Napari debugging during plugin development documentation

seankmartin opened the pull request:
# Description
This PR adds a page to the the napari.org/plugins category about debugging plugins during development.

## Type of change
- [x] Documentation update

# References
closes #2948 closes #3735

Forum references:
* [Tracebacks on Image.sc](https://forum.image.sc/t/do-not-see-tracebacks-when-using-magicgui-to-create-widgets-in-napari/57861/2).
* [Reloading changed code on Image.sc](https://forum.image.sc/t/debug-develop-napari-plugin-how-to-reload-changed-code/60618/4).
* [Plugging logging on zulipchat](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/logging.20for.20plugins)

[Here is the code used in the docs](https://github.com/seankmartin/napari-plugin-debug)

# How has this been tested?
* I manually checked that the code in the docs runs as expected.
* `make docs` runs successfully and builds the docs as expected.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

seankmartin committed:
Add debugging plugin docs

seankmartin committed:
Small debug doc changes and add to TOC

seankmartin committed:
Clean up debug docs

seankmartin committed:
Small debug updates

DragaDoncila commented on docs/plugins/debug_plugins.md:
````diff
@@ -0,0 +1,365 @@
+```{note}
+In general, `napari --info` is a good first step to debugging any environment issues and providing the output from this command is useful when raising bugs.
+```
+
````
I would add a note here to make sure that 

1. your plugin package is visible when you run `pip list` or `conda list`
2. your entry point is configured correctly as per https://napari.org/stable/plugins/first_plugin.html#update-setup-cfg

For example if `pip list` *does* show the package is installed, and `npe2 validate` reports success, but `napari` can't see your plugin, your entry point may be ill-configured.

DragaDoncila commented on docs/plugins/debug_plugins.md:
````diff
@@ -0,0 +1,365 @@
+In general, `napari --info` is a good first step to debugging any environment issues and providing the output from this command is useful when raising bugs.
+```
+
+## Seeing plugin errors and warnings
````
I feel we may want to remove this section entirely. Afaik we generally consider the `Plugin Errors` dialog a poor decision and want to get rid of it, so probs best not to mention it in the docs. I've never had a plugin error not reported to the command line so I would just recommend users inspect that (as below).

seankmartin replied:
Done!

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,365 @@
+Inside of this popup, the full traceback can be seen, along with the option to drop into the debugger from here.
+Dropping into the debugger will open the built in [python debugger](https://docs.python.org/3/library/pdb.html) at the point of failure.
+
+You can also set napari not to catch error messages, or set napari exit on error via the following environment variables, respectively:
```
```suggestion
You can also configure napari not to catch error messages, or force napari to exit on error via the following environment variables, respectively:
```

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,365 @@
+
+### Isolate the plugin functionality from napari
+
+Since napari plugins are just Python functions, it can be easiest to isolate the issues from napari in some circumstances. For example, in the case of a widget, the widget could be isolated from napari and run separately as a PyQt widget for debugging purposes. In this way, the usual debugging tools, such as those provided by an IDE, can be applied. This method has the additional benefit of align ...
```
```suggestion
To simplify the debugging process, it can be useful to isolate the execution of your plugin code from executing napari itself. For example, in the case of a widget, the widget could be isolated from napari and run separately as a PyQt widget for debugging purposes. In this way, the usual debugging tools, such as those provided by an IDE, can be applied. This method has the additional benefit of aligning with test driven development.
```

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,365 @@
+### Write a script to setup the plugin without user interaction
+
+Writing a setup script reproducibly places your napari viewer at the state of problems, without having to perform user interaction in the GUI to get back to the problem state.
+
```
I would add a note here that launching napari & your plugin from a script can often allow you to use your IDE's built in debugging tools, as long as both your plugin and napari itself are installed in editable mode in your environment.

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,365 @@
+
+## A simple plugin for following examples
+
+To demonstrate these debugging options, we will create a small plugin that prints the text entered into a box by the user. The plugin will provide an autogenerated widget, a `magic_factory` widget, and a `QWidget` - see the [contribution guides](./guides.md#widgets) for more detail. Below is the plugin file structure and contents. Create this structure, and fill the files with the below informati ...
```
It would be nice if we could provide this sample project in a repo the user can clone rather than having them create each file manually?

seankmartin replied:
Yes, I thought this too but I was not sure what the practices for linking napari docs to other repos, so I wrote the code in the doc.
I do have a repository [here](https://github.com/seankmartin/napari-plugin-debug) with the code that I will suggest cloning for now.

DragaDoncila commented on docs/plugins/debug_plugins.md:
````diff
@@ -0,0 +1,365 @@
+* = *.yaml
+```
+
+Then install the package by:
````
```suggestion
Then install the package into an environment already containing `napari` by:
```

DragaDoncila commented on docs/plugins/debug_plugins.md:
````diff
@@ -0,0 +1,365 @@
+    return output_string
+```
+
+Notice how the widgets in the viewer currently have the same behaviour as before, despite the code update. This is because the code won't immediately update, as IPython does not yet know that it should update that code. To trigger the code update, just type the name of your changed function into IPython, or reference it in some way:
````
```suggestion
To see the updated behavior in action, just type the name of your changed function into IPython, or reference it in some way:
```

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,365 @@
+
+### Changing the code with IPython running
+
+Currently, clicking on the run button for any of these widgets in the napari viewer outputs "You entered **YOUR_ENTRY**!". However, we would like to change the behaviour such that a special message is printed if nothing is entered. So let's change our `example` function to:
```
```suggestion
Currently, clicking on the run button for any of these widgets in the napari viewer outputs "You entered **YOUR_ENTRY**!". However, we would like to change the behaviour such that a special message is printed if nothing is entered. So let's change our `example` function in `_widget.py` to:
```

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,365 @@
+
+## Isolate the issue from napari
+
+This solution ties in with the idea of test-driven development (see the [napari testing guidelines](./test_deploy.md#prefer-smaller-unit-tests-when-possible)). The idea is to trust that napari will provide the information you expect it to, and test your widgets independently of the viewer. In the case above, to verify that some input values work as expected, it would be like so:
```
```suggestion
This solution ties in with the idea of test-driven development (see the [napari testing guidelines](./test_deploy.md#prefer-smaller-unit-tests-when-possible)). The idea is to trust that napari will provide the information you expect it to, and test your widgets independently of the viewer. In the case above we can verify that input values work as expected like so:
```

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,365 @@
+There are, generally speaking, three main methods for notifying users of problems in napari.
+
+1. Raise an exception to indicate a breaking problem in the code (e.g. unexpected user input `raise ValueError("some error")`).
+2. Indicate that something that was handled, but may not be the behaviour the user was expecting using `warnings.warn("some warning")`.
```
```suggestion
2. Indicate that something was handled, but may not be the behaviour the user was expecting using `warnings.warn("some warning")`.
```

DragaDoncila reviewed (changes requested):
This is looking great @seankmartin, thank you so much for writing such a detailed doc 🎉 ! I've left a couple of small tweaky comments and a couple just requesting a few notes added.

Let me know if I can clarify anything!

seankmartin committed:
Apply suggestions from code review

Co-authored-by: Draga Doncila Pop <17995243+DragaDoncila@users.noreply.github.com>

seankmartin committed:
Incorporate review comments

- Remove section about warning menu
- Add point about checking for entry point config
- Add point about debugging with IDE
- Link to code repo instead of inline code for starting point

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,290 @@
+You should also ensure that:
+
+1. Your plugin package is visible when you run `pip list` or `conda list` (depending on your environment setup).
+2. Your entry point is correctly configured, see the [first steps](./first_plugin.md#update-setup-cfg) for information.
```
```suggestion
2. Your entry point is correctly configured, see the [first steps](./first_plugin#update-setup-cfg) for information.
```

DragaDoncila committed:
Update docs/plugins/debug_plugins.md

DragaDoncila commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,290 @@
+You should also ensure that:
+
+1. Your plugin package is visible when you run `pip list` or `conda list` (depending on your environment setup).
+2. Your entry point is correctly configured, see the [first steps](./first_plugin#update-setup-cfg) for information.
```
```suggestion
2. Your entry point is correctly configured, see the [first steps](first_plugin.md#update-setup-cfg) for information.
```

DragaDoncila committed:
Update docs/plugins/debug_plugins.md

DragaDoncila committed:
Merge branch 'main' into napari-debug-docs

melissawm commented on docs/plugins/debug_plugins.md:
````diff
@@ -0,0 +1,290 @@
+    test_false_inputs()
+```
+
+Then, for `python test_print.py` you can use any of your usual debugging tools - such as the visual debugger provided by a Python IDE (e.g. PyCharm, VSCode, or Spyder). Further, an isolated test like this can be integrated into a [testing suite for your napari plugin](https://napari.org/stable/plugins/test_deploy.html).
````
Whenever possible, it is better to use internal links instead of linking to the actual napari.org website.

```suggestion
Then, for `python test_print.py` you can use any of your usual debugging tools - such as the visual debugger provided by a Python IDE (e.g. PyCharm, VSCode, or Spyder). Further, an isolated test like this can be integrated into a [testing suite for your napari plugin](test_deploy).
```

melissawm commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,290 @@
+
+### Set up plugin log messages
+
+In addition to these user focused methods, you can set up plugin debug logs and messages during development. You can either use [napari specific functions](https://napari.org/dev/api/napari.utils.notifications.html), or [built in Python logging](https://docs.python.org/3/library/logging.html).
```
Here also we can use internal links. The syntax is from [MyST](https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html#targets-and-cross-referencing)

```suggestion
In addition to these user focused methods, you can set up plugin debug logs and messages during development. You can either use {mod}`napari specific functions <napari.utils.notifications>`, or [built in Python logging](https://docs.python.org/3/library/logging.html).
```

melissawm commented on docs/plugins/debug_plugins.md:
```diff
@@ -0,0 +1,290 @@
+You should also ensure that:
+
+1. Your plugin package is visible when you run `pip list` or `conda list` (depending on your environment setup).
+2. Your entry point is correctly configured, see the [first steps](first_plugin.md#update-setup-cfg) for information.
```
This is an interesting one - the [autogenerated heading anchors](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#syntax-header-anchors) may behave differently with different syntax (in this case, the target section is called ``Update `setup.cfg` ``, and the dot is removed when creating the link.

Here's a tip of how to inspect those generated links:

```
(napari-dev) ~/napari (napari-debug-docs ✗) myst-anchors -l 3 docs/plugins/first_plugin.md 
<h1 id="your-first-plugin"></h1>
<h2 id="before-you-start"></h2>
<h2 id="what-is-a-plugin"></h2>
<h2 id="1-create-a-new-directory"></h2>
<h2 id="2-add-standard-python-package-metadata-files"></h2>
<h3 id="pyprojecttoml"></h3>
<h3 id="setupcfg"></h3>
<h2 id="3-implement-the-plugin"></h2>
<h3 id="add-plugin-functionality-to-napari_hello__init__py"></h3>
<h3 id="add-a-napariyaml-manifest"></h3>
<h3 id="update-setupcfg"></h3>
<h2 id="4-install-your-plugin-and-try-it-out"></h2>
<h2 id="get-going-quickly-with-cookiecutter"></h2>
<h2 id="next-steps"></h2>
```

melissawm replied:
```suggestion
2. Your entry point is correctly configured, see the [first steps](first_plugin.md#update-setupcfg) for information.
```

melissawm commented on docs/plugins/debug_plugins.md:
````diff
@@ -0,0 +1,290 @@
+
+# Now the code is reloaded, run outputs as expected
+Please enter something in the text box.
+```
````
This code block is generating errors when compiling - I am guessing this comes from the `!` character in the code block. The following works, but I'll leave it as a suggestion, maybe there are better ones:

```suggestion
```IPython
# Run before changing the code
'You entered !'

# Run after changing the code, but before triggering an update
'You entered !'

In [10]: example # anything in IPython to reload that code

# Now the code is reloaded, run outputs as expected
Please enter something in the text box.
```
```

melissawm commented on docs/plugins/debug_plugins.md:
````diff
@@ -0,0 +1,290 @@
+
+Running this script with `python example_notification.py` and entering fast into the input text box and clicking run you should then see:
+
+```txt
````
This is also causing errors with the MyST parser:

```suggestion
```
```

melissawm reviewed (commented):
I left a few comments, hope those are useful - let me know if I can help further!

seankmartin committed:
Apply suggestions from code review

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

seankmartin committed:
Fix myst parser errors
